### DISCUSSION TITLE
defining schema for a schema

### DISCUSSION THREAD

**Discussion by rashmy-git** [json-schema-org/community#736] [Q&A]
Hi,
I have previously used simple features of JSON schema to validate my JSON data.
Now, I'm exploring the option of defining a JSON schema to validate another JSON schema.

In my service, I want to allow the user to create any kind of business object. Say the example of blog post from [here](https://json-schema.org/learn/json-schema-examples#blog-post).
```code
POST/create-data

{
  "title": "New Blog Post",
  "content": "This is the content of the blog post...",
  "publishedDate": "2023-08-25T15:00:00Z",
  "author": {
    "username": "authoruser",
    "email": "author@example.com"
  },
  "tags": ["Technology", "Programming"]
}
```
But before any data is created, I'd like my user to give me the JSON schema for their business object.  So the JSON schema for the above data would look like this:
```code
POST /define-schema

{
  "$id": "https://example.com/blog-post.schema.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "description": "A representation of a blog post",
  "type": "object",
  "required": ["title", "content", "author"],
  "properties": {
    "title": {
      "type": "string"
    },
    "content": {
      "type": "string"
    },
    "publishedDate": {
      "type": "string",
      "format": "date-time"
    },
    "author": {
      "$ref": "https://example.com/user-profile.schema.json"
    },
    "tags": {
      "type": "array",
      "items": {
        "type": "string"
      }
    }
  }
}
```

Now when they give me the schema, how can I allow the user to include the below in the JSON schema?
1. A sibling of "properties" called "deploy-region" with keys "primary" and "secondary", where primary has enum [us-east, us-west]
2. Can I extend "type" to have additional values like "businessObjectType" apart from the standard values [object, string, number, boolean etc..]

Basically, I want to create a schema for a schema.

With my limited search, I could not find good answers. If there is any documentation or previous references please point me to them.

Thank you!

**Comment by jdesrosiers** [comment]
A schema that validates a schema is known as a "meta-schema". Meta-schemas are just schemas. They can be a little mind bending to write, but you can write them just like you would write any other schema.

For example, here's a constraint that requires that there's `properties` keyword has a `deploy-region` property.

```json
{
  "properties": {
    "properties": {
      "required": ["deploy-region"]
    }
  }
}
```

`properties` in this example shows how things can get confusing when writing meta-schemas.

So, using my example, see if you can expand it to express the rest of your constraints. If you're still having trouble, post what you've tried and we'll try to nudge you in the right direction.

Also, sorry it took so long for someone to respond. Our whole team was traveling last week and we've gotten a bit behind.

**Comment by rashmy-git** [comment]
Thank you for the reply!

I'm trying to wrap my head around vocabularies and connecting it with dialect :)
I've made an attempt at it, but the problem is that when using it ajv to validate, it fails.

Here is what I have:

### Example JSON Instance Data
```json
{
	"deployapi": "1.0",
	"title": "class-of-2024-blogs",
	"blogs": {
		"blog-schemas": {
		  "ScienceBlogs": {
			"type": "object",
			"properties": {
			  "topic": {
				"type": "string",        //REUSE all fields available in JSON schema 'properties'
				"minLength": 10,
				"maxLength": 128,
				"pattern": "^[a-zA-Z0-9]*$",
				"example": "Dark Matter",
				"primary-key": true      //NEW field, that should be available within properties
			  },
			  "writers": {
				"type": "array",
				"items": {
				  "type": "string"
				},
				"example": ["Jack", "Jill"]
			  },
			  "content": {
				"type": "string",
			  },
			  "publish-date": {
				"type": "string",
				"format": "date-time",
			  }
			},
			"required": ["topic", "writers", "content", "publish-date"],
			"additionalProperties": false
			"deploy-region": "us-east",   //NEW field, that should be available as sibling to properties
			"owner-contact": {            //NEW field, that should be available as sibling to properties
				"email": "test@gmail.com"
			}
		  },
		  "TheatreBlogs": {
			"type": "object",
			"properties": {
			  "title": {
				"type": "string",
				"primary-key": true
			  },
			  "writers": {
				"type": "array",
				"items": {
				  "type": "string"
				}
			  },
			  "actors": {
				"type": "array",
				"items": {
				  "type": "string"
				}
			  },
			  "genre": {
				"type": "string",
				"enum": ["comedy", "historical"]
			  }
			  "synopsis": {
				"type": "string",
			  },
			  "publish-date": {
				"type": "string",
				"format": "date-time",
			  }
			},
			"required": ["topic", "writers", "actors", "synopsis", "publish-date"],
			"additionalProperties": false
			"deploy-region": "us-west",
			"owner-contact": {
				"email": "drama@gmail.com"
			}
		  },
		}
	}
}
```

### My schema
```json
{
	"$schema": "https://myorganization/meta-schema#",
	"$id": "https://myorganization/blog-schema##",


	"type": "object",
	"properties": {
		"deployapi": {
		  "type": "string",
		  "pattern": "^1\\.\\d+(-.+)?$"
		},
		"title": {
		  "type": "string"
		},
		"blogs": {
		  "$ref": "#/$defs/blogModel"
		},
		"required": [ "deployapi", "title", "blogs"]
		"additionalProperties": false
	}
	"$defs": {
		"blogModel": {
			"type": "object",
			"properties": {
				"blog-schemas": {
				  "type": "object",
				  "properties": {
					"deploy-region": {
					  "$ref": "#/$defs/deployRegion"
					},
					"owning-team": {
					  "$ref": "#/$defs/owningTeam"
					}
				  }
 				  "required": ["deploy-region", "owning-team"],
				  "additionalProperties": {
					"$dynamicRef": "#meta"
				  }
				}
			}
		}
		"deployRegion": {
		   "type": "string"
		   "enum": ["us-east", "us-west"]
		},
		"owningTeam": {
			"type": "object",
			"properties": {
				"contact-name": {
					"type": "string"
				},
				"contact-email": {
					"type": "string"
				}
			},
			"required": ["contact-email"],
			"additionalProperties": false
		},
	}
}
```

### My meta-schema
```json
{
	"$schema": "https://myorganization/meta-schema#", 
	"$id": "https://myorganization/meta-schema#",
	"$vocabulary": {
		"https://json-schema.org/draft/2020-12/vocab/core": true,
		"https://json-schema.org/draft/2020-12/vocab/applicator": true,
		"https://json-schema.org/draft/2020-12/vocab/validation": true,
		"https://json-schema.org/draft/2020-12/vocab/meta-data": true,
		"https://json-schema.org/draft/2020-12/vocab/format-annotation": true,
		"https://json-schema.org/draft/2020-12/vocab/content": true,
		"https://myorganization/myorganization-vocab": true
	},
	"$dynamicAnchor": "meta",

	"type": ["object", "boolean"],
	"properties": {
		"example": true,
		"primary-key": {
		   "type": "boolean"
		}
	}
}
```

**Questions:**
[1]
When using AJV code to validate, with the above schema and JSON, it is failing.
```
ajv.addMetaSchema( /**Added My meta-schema **/ )
const validate = ajv.compile( /**Provide My schema **/ )
validate( /**Provide Example JSON Instance data **/ )
```
	
[2]
Do I need to use dialect as well?
If there is dialect, vocabulary, meta-schema how do I provide these to ajv?
Should I use bundling to generate a single meta-schema and then provide this to ajv validator?

Thanks for your time and guidance!

**Reply by jdesrosiers** [reply]
It looks like you have several errors in your schema. Any one of them could be causing you issue with ajv. Ajv doesn't support `$vocabulary`, but that probably doesn't matter in your case. All you really need is the custom meta-schema.

Here are the issues I found.
* You are missing commas in several places. Use a JSON validator to find these.
* `required` and `additionalProperties` are in the wrong place in the top of your schema. They should be after `properties` not inside it. 
* You're using `"$dynamicRef": "#meta"` without a `$dynamicAnchor`. What do you expect this to resolve to?

**Reply by jdesrosiers** [reply]
It looks like you don't want a dynamic reference at all. I think this is what you want.

```json
"blogModel": {
  "type": "object",
  "properties": {
    "blog-schemas": {
      "type": "object",
      "additionalProperties": {
        "type": "object",
        "properties": {
          "deploy-region": {
            "$ref": "#/$defs/deployRegion"
          },
          "owning-team": {
            "$ref": "#/$defs/owningTeam"
          }
        },
        "required": ["deploy-region", "owning-team"]
      }
    }
  }
}
```

**Reply by jdesrosiers** [reply]
Actually, probably this,

```json
"blogModel": {
  "type": "object",
  "properties": {
    "blog-schemas": {
      "type": "object",
      "additionalProperties": {
        "$ref": "https://myorganization/meta-schema",
        "properties": {
          "deploy-region": {
            "$ref": "#/$defs/deployRegion"
          },
          "owning-team": {
            "$ref": "#/$defs/owningTeam"
          }
        },
        "required": ["deploy-region", "owning-team"]
      }
    }
  }
}
```

**Reply by rashmy-git** [reply]
Thank you for all your suggestions. I'll use the latest one you suggested and test out my scenarios.

Sorry for the late reply, I was out.
